File: typescript/src/ancilis/middleware/action-builder.ts
/** Translates MCP tool calls into framework-agnostic Action objects. */

import { createHash, randomUUID } from "node:crypto";
import type { ResolvedConfig } from "../config/index.js";
import type { Action } from "../engine/action.js";
import type { ToolRegistry } from "../engine/registry.js";

export function buildAction(
  toolName: string,
  args: Record<string, unknown> | undefined,
  config: ResolvedConfig,
  registry: ToolRegistry,
): Action {
  const rawArgs = args ?? {};
  const paramHash = createHash("sha256")
    .update(JSON.stringify(rawArgs, Object.keys(rawArgs).sort()))
    .digest("hex");

  const entry = registry.lookup(toolName);
  const toolVersion = entry?.version ?? null;
  const descriptionHash = entry?.descriptionHash ?? null;

  const dcCodes: string[] = [];
  for (const codes of config.dataClassifications.values()) {
    for (const code of codes) {
      if (!dcCodes.includes(code)) dcCodes.push(code);
    }
  }

  return {
    actionId: randomUUID(),
    timestamp: new Date().toISOString(),
    agentId: config.agentName,
    agentOwner: config.agentOwner || null,
    actionType: "tool_call",
    tool: {
      name: toolName,
      version: toolVersion,
      descriptionHash,
    },
    parameters: { raw: rawArgs, parameterHash: paramHash },
    context: {
      dataClassifications: dcCodes,
      activeOverlays: [...config.activeOverlays.keys()],
    },
  };
}

File: python/src/ancilis/middleware/action_builder.py
"""Translates MCP tool calls into framework-agnostic Action objects."""

from __future__ import annotations

import hashlib
import json
import uuid
from datetime import datetime, timezone
from typing import Any

from ancilis.config import ResolvedConfig
from ancilis.engine.action import Action, ActionContext, ActionParameters, ToolInfo
from ancilis.engine.registry import ToolRegistry


def build_action(
    tool_name: str,
    arguments: dict[str, Any] | None,
    config: ResolvedConfig,
    registry: ToolRegistry,
) -> Action:
    """Build an Action object from an MCP tool call."""
    args = arguments or {}
    param_hash = hashlib.sha256(json.dumps(args, sort_keys=True, default=str).encode()).hexdigest()

    # Look up tool metadata from registry
    entry = registry.lookup(tool_name)
    tool_version = entry.version if entry else None
    description_hash = entry.description_hash if entry else None

    # Collect DC codes from config
    dc_codes: list[str] = []
    for codes in config.data_classifications.values():
        for code in codes:
            if code not in dc_codes:
                dc_codes.append(code)

    overlay_ids = list(config.active_overlays.keys())

    return Action(
        action_id=str(uuid.uuid4()),
        timestamp=datetime.now(timezone.utc).isoformat(),
        agent_id=config.agent_name,
        source_type="mcp",
        agent_owner=config.agent_owner or None,
        action_type="tool_call",
        tool=ToolInfo(
            name=tool_name,
            version=tool_version,
            description_hash=description_hash,
        ),
        parameters=ActionParameters(raw=args, parameter_hash=param_hash),
        context=ActionContext(
            data_classifications=dc_codes,
            active_overlays=overlay_ids,
        ),
        producer_type="mcp",
    )
